services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java

```diff
@@ -456,28 +457,15 @@ private void validateTokenReuseForRefresh(KeycloakSession session, RealmModel re
     }
 
     // Will throw OAuthErrorException if validation fails
-    private void validateTokenReuse(KeycloakSession session, RealmModel realm, AccessToken refreshToken,
-        AuthenticatedClientSessionModel clientSession, boolean refreshFlag) throws OAuthErrorException {
-        int startupTime = session.getProvider(UserSessionProvider.class).getStartupTime(realm);
-
-        if (clientSession.getCurrentRefreshToken() != null
-            && !refreshToken.getId().equals(clientSession.getCurrentRefreshToken())
-            && refreshToken.getIat() < clientSession.getTimestamp()
-            && startupTime <= clientSession.getTimestamp()) {
-            throw new OAuthErrorException(OAuthErrorException.INVALID_GRANT, "Stale token");
-        }
+    private void validateTokenReuse(RealmModel realm, AccessToken refreshToken, AuthenticatedClientSessionModel clientSession) throws OAuthErrorException {
+        int lastRefresh = clientSession.getRefreshTokenLastRefresh(refreshToken.getId());
```

Comment:
Using `iat` allows the refresh token to be used only for X seconds (Reuse Period) after `iat`, but the intent is to allow the refresh token to be reused immediately after its refresh for X seconds (Reuse Period). 
In other words, you should start counting the reuse period from the moment of the refresh and not from the moment of issuance (`iat`) of the refresh token. 
An improvement could be to save the last refresh only when the maximum reuse limit is reached to avoid potential infinite reuse of the same refresh token in case it is refreshed during the "reuse period" just after refresh.

Reply:
Ah! OK, so it's the time of the new refresh, not the previous one. Understood. Please resolve this conversation (I cannot do it, I suppose I have no permissions).